mitiq/qse/qse_utils.py

```diff
@@ -0,0 +1,134 @@
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+"""Functions for computing the projector for subspace expansion."""
+
+from typing import Callable, Sequence, Union
+from mitiq import Observable, QPROGRAM, QuantumResult, PauliString
+from typing import Dict
+import numpy as np
+import numpy.typing as npt
+from scipy.linalg import eigh
+from numpy.linalg import pinv
+
+
+def get_projector(
+    circuit: QPROGRAM,
+    executor: Callable[[QPROGRAM], QuantumResult],
+    check_operators: Sequence[PauliString],
+    code_hamiltonian: Observable,
+) -> Observable:
+    """Computes the projector onto the code space defined by the
+    check_operators provided that minimizes the code_hamiltonian.
+
+    Returns: Projector as an Observable.
+    """
+    pauli_string_to_expectation_cache: Dict[PauliString, complex] = {}
+    S = _compute_overlap_matrix(
+        circuit, executor, check_operators, pauli_string_to_expectation_cache
+    )
+    H = _compute_hamiltonian_overlap_matrix(
+        circuit,
+        executor,
+        check_operators,
+        code_hamiltonian,
+        pauli_string_to_expectation_cache,
+    )
+    # We only want the smallest eigenvalue and corresponding eigenvector
+    _, C = eigh(pinv(S) @ H, subset_by_index=[0, 0])
+    # np float type: np.float64
+
+    Cs = C[:, 0]
+    projector = Observable(
+        *[check_operators[i] * Cs[i] for i in range(len(check_operators))]
+    )
+
+    return projector
+
+
+def _get_expectation_value_for_observable(
+    circuit: QPROGRAM,
+    executor: Callable[[QPROGRAM], QuantumResult],
+    observable: Union[PauliString, Observable],
+    pauli_string_to_expectation_cache: Dict[PauliString, complex] = {},
+) -> float:
+    def get_expectation_value_for_one_pauli(
+        pauli_string: PauliString,
+    ) -> float:
+        cache_key = pauli_string.with_coeff(1)
+        pauli_string_to_expectation_cache[cache_key] = Observable(
+            cache_key
+        ).expectation(circuit, executor)
+        return (
+            pauli_string_to_expectation_cache[cache_key] * pauli_string.coeff
+        ).real
+
+    total_expectation_value_for_observable = 0.0
+
+    if isinstance(observable, PauliString):
+        pauli_string = observable
+        total_expectation_value_for_observable += (
+            get_expectation_value_for_one_pauli(pauli_string)
+        )
+    elif isinstance(observable, Observable):
+        for pauli_string in observable.paulis:
+            total_expectation_value_for_observable += (
+                get_expectation_value_for_one_pauli(pauli_string)
```

Comment:
That's a good point. I didn't realize that `Executor.evaluate(Observable)` was doing QWC grouping.

I'm not sure if we need to choose between caching and grouping compatible terms. I'm assuming `Executor.evaluate` would be computing each term and adding them up at the end (is this true?). Can I introduce caching at the step right before the addition?

Assuming we need to choose between caching and grouping QWC terms, assuming the user provided a set of stabilizers that form a group, caching allows us to measure only 2^m times instead of 2^2m times (where m is the number of check operators). I am not sure how many measurements QWC grouping would save us, will think about this some more.

Reply:
> That's a good point. I didn't realize that `Executor.evaluate(Observable)` was doing QWC grouping.
> 
> I'm not sure if we need to choose between caching and grouping compatible terms. I'm assuming `Executor.evaluate` would be computing each term and adding them up at the end (is this true?). Can I introduce caching at the step right before the addition?

If there is a way to have both caching and grouping, that would be optimal. I actually don't see an obvious way right now, but maybe you have an idea.  In the worst case, the current code with caching is totally ok.

> Assuming we need to choose between caching and grouping QWC terms, assuming the user provided a set of stabilizers that form a group, caching allows us to measure only 2^m times instead of 2^2m times (where m is the number of check operators). I am not sure how many measurements QWC grouping would save us, will think about this some more.

I think in this context the advantage of caching is much more important. Grouping has a significant advantage  when you have a sum of many local observables, while here we are dealing with long and generic Pauli strings which are unlikely to be measurable simultaneously. So, caching seems much more important in this setting.

Reply:
> I actually don't see an obvious way right now, but maybe you have an idea. In the worst case, the current code with caching is totally ok.
I just took a look at the code, it seems a bit more complicated than what I thought. I might work on it in another PR in the future.

Just brainstorming here:
I was thinking a simple approach could have been caching the `pset`s, but that's not very useful because they are currently generated randomly. So maybe generating them determinstically could help? But what would the downside be for generating them determinstically?

>  while here we are dealing with long and generic Pauli strings which are unlikely to be measurable simultaneously.

Since we are dealing with stabilizer codes, we expect all the pauli strings to measured simulatenously, although they are not necessarily QWC as you mentioned. Something like [this](https://pubs.acs.org/doi/10.1021/acs.jctc.0c00008) could be employed, although this would require an exponential complexity for classical processing.

Reply:
I think the best thing is to keep the current code, i.e, keeping the sequential measurement of Pauli strings with caching. Maybe there is some space for further optimizations but they would require too many changes in the way the Executor class works. I don't think it's a good idea to try it in this PR. 

So, given that I don't have extra comments (thanks for adding comments on the tests), we can proceed and merge this. Thanks @bubakazouba !